### ISSUE TITLE
Handling duplicate trips in source GTFS data

### ISSUE THREAD

**Issue by seanblanchfield** [seanrees/gtfs-upcoming#7]
I sometimes see duplicate buses in my feed, apparently due to mistakes in the GTFS schedule from TFI. Consider the following example:
``` json
{
            "trip_id": "3531_6248",
            "route": "27",
            "route_type": "BUS",
            "headsign": "Jobstown",
            "direction": "0",
            "stop_id": "8220DB001409",
            "dueTime": "13:43:28",
            "dueInSeconds": 237.596308,
            "source": "SCHEDULE"
        },
        {
            "trip_id": "3467_6234",
            "route": "27",
            "route_type": "BUS",
            "headsign": "Jobstown",
            "direction": "0",
            "stop_id": "8220DB001409",
            "dueTime": "13:43:28",
            "dueInSeconds": 237.597341,
            "source": "SCHEDULE"
        },
```

I investigated further, and I see the following corresponding entries in `trips.txt`:
``` csv
route_id,service_id,trip_id,trip_headsign,trip_short_name,direction_id,block_id,shape_id
...
3467_50384,358,3531_6248,Jobstown,5999,0,27006,3531_179
3467_50384, 76,3467_6234,Jobstown,5999,0,27006,3467_179
```
The trips differ only in the `service_id`, however when I find the corresponding entries in `calendar.txt` I see the services don't have any meaningful difference:
``` csv
service_id,monday,tuesday,wednesday,thursday,friday,saturday,sunday,start_date,end_date
358,1,1,1,1,1,0,0,20230828,20231201
 76,1,1,1,1,1,0,0,20230821,20230825
```
I would like to discuss how to deduplicate trips like this. I am willing to implement a fix, but I'd like to take your feedback first. 

Suggestion: replace the `service_id` in the trip record with the corresponding calendar entry, and then consider them duplicates of the resulting data is identical (same route, same calendar etc).

Thoughts?

**Comment by seanblanchfield** [comment]
On closer examination I do in fact see a significant difference between the service entries in `calendar.txt` above: Their start/end dates define periods that do not overlap (although they are nearly consecutive).  I think the problem is that the code is not checking the scheduled service's start/end dates. Perhaps this check belongs in database.py:IsValidServiceDay()?

On a related note, I can't find any NTA documentation that says when the data in `calendar.txt` gets updated, but from looking at the entries I suspect i changes every Sunday. To keep things accurate, I suppose it would be necessary to redownload the zip file every week and restart the server.  If this concurs with your understanding, perhaps it should be mentioned in the README.

**Comment by seanrees** [comment]
I think you're right. `_IsValidServiceDay` seems an appropriate place. At no point are the `start_date` and `end_date` of the calendar used (and on line 234 is a TODO :))

**Comment by seanblanchfield** [comment]
You can pop the following code in at `database.py:116` to fix this:

``` python
    start_date = datetime.datetime.strptime(service['start_date'], "%Y%m%d").date()
    end_date = datetime.datetime.strptime(service['end_date'], "%Y%m%d").date()
    if dt < start_date or dt > end_date:
      return False
```

One implication of this is that if the GTFS zip file is more than a week old, a lot of services will be discarded in this way. To get the new corresponding "in-date" services, you need to re-download the zip file, extract it and restart the server.

**Comment by seanrees** [comment]
That seems straightforward enough, thanks! Will look to make this change in a few days after some biz travel.

**Comment by seanrees** [comment]
This is now fixed in #8 (thanks for waiting -- life got pretty seriously in the way)
